### PR TITLE
chat: federation, split hook into pull- and push-hook

### PR DESCRIPTION
**Pull request by pilfer-pandex**
This change provides for the federation of chat channels. This means that a ship which does not own a channel but does subscribe to it can (if it opts in) be subscribed to in turn. Additionally, a mechanism for a ship to block some amount of incoming watch requests for a chat and instead "redirect" them to other ships is provided. These capabilities operate orthogonally.

Suppose ~zod hosts ~/~zod/chat. To add ~nec as a proxy for this channel:
1. Make sure ~nec's chat is connected to it. E.g. in ~nec's chat-cli, run `;join ~/~zod/chat`
2. On ~zod, run `:chat-push-hook &chat-push-hook-action [%set-proxies /~/~zod/chat (sy ~nec ~)]`
3. On ~nec, run  `:chat-push-hook &chat-push-hook-action [%add /~/~zod/chat &]` (or `|` to disallow the syncing of chat histories)
4. Now when a foreign ship connects to ~zod, she may instead be asked to connect to ~nec, depending on whether her address is even or odd.

This pr also implements a major architecting of the chat replication subsystem:

Before this change, the chat backend was divided in two parts: the chat-store, which maintained the "database" of chats, and the chat-hook, which provided for replication of this database across ships. This change splits the chat-hook into two parts: the chat-pull-hook and the chat-push-hook.

Suppose ~zod hosts a channel foo and ~nec subscribes to it. ~zod's chat-store has the authoritative copy of the history of that channel, and it is to ~zod's store that all writes must go. When ~zod sends a message, it writes to its store. ~zod's push-hook is watching that store and sends an update to all of its subscribers on the path /mailbox/~zod/foo with the new message. As a client of this chat, ~nec has set up its pull-hook as one of these subscribers. When it receives the fact containing the new message, it writes that message to its own store. In this rudimentary way, it is effected that ~nec's accounting of ~zod/foo in its chat-store is a dynamically updating copy of ~zod's.

When ~nec wants to send a message, it sends it to ~zod's chat-push-hook, which does a permission check before forwarding to ~zod's store. The rest of the flow proceeds as before.

All of this behavior was previously handled in a single chat-hook app, which was accordingly extremely hard to follow, and had a lot of event spaghetti and punning of messages. As a result of this change, suggested by @loganallenc, it should be much more straightforward (indeed much simpler than this change) for me to implement a chat proxying system.

### CODE REVIEW

File: pkg/arvo/lib/chat.hoon [hunk 1]
```diff
@@ -0,0 +1,10 @@
+^?
+|%
+++  ship-from-path
```
Comment (Fang-) [inline pkg/arvo/lib/chat.hoon:3]: You might consider putting this into `/lib/chat-store`, too. It isn't doing anything generic with chat data-types, and only implements some logic wrt how chat-store deals with paths.

(Of course, right now, we do this in multiple places, so a `/lib/userspace` or w/e might be justified...)

Comment (ohAitch) [reply pkg/arvo/lib/chat.hoon:3]: \> /lib/misc

**Comment by pilfer-pandex** [comment]
Yeah @Fang-, so I disagree with you on three letter names. Having experimented with both, I feel that using the autonamer and shadowing the type definition (so that e.g. "path" refers to a specific path rather than the type of all paths) is a bad pattern. Autonamer ok in e.g. $%, but not when it dumps the result directly into the namespace.

Definitely disagree with you on =/.

**Comment by philipcmonk** [comment]
Autonaming is definitely good in definitions where it doesn't go into the global namespace, and less good otherwise but often still worth it.  We haven't banned 3-letter names, but they're becoming less common.  I have reservations but have usually found them worth it.  Definitely they should be canonical and pronounceable though, so eg `our`, `her`, or `him` (for 1st person, 2nd person, and 3rd person respectively) instead of `shp` (what's canonical for each type?  You either know it because you've memorized the entire codebase or you start to see why using the type name as a the canonical name is convenient).  And generally try to match the style of what's around if it's small changes to existing code.

IMO `=/` is fine with or without a type.  If you find yourself wondering what the type is, just add it as useful documentation, but it's not necessary.

File: pkg/arvo/app/chat-pull-hook.hoon [hunk 2]
```diff
@@ -0,0 +1,260 @@
+/-  *invite-store, hook=chat-pull-hook
+/+  default-agent, verb, dbug, store=chat-store, lib=chat-hooks
+|%
++$  card  card:agent:gall
++$  versioned-state
+  $%  [%0 state-0]
+  ==
++$  state-0
+  $:  tracking=(map path ship)
```
Comment (loganallenc) [inline pkg/arvo/app/chat-pull-hook.hoon:9]: Since you’re starting this hook from scratch anyway, could you make this a `(map [=ship =term] ship)` ?

File: pkg/arvo/app/chat-pull-hook.hoon [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+--
+::
+=|  [%0 state-0]
+=*  state  -
+::
+%-  agent:dbug
+%+  verb  |
+::
+^-  agent:gall
+|_  bol=bowl:gall
```
Comment (loganallenc) [inline pkg/arvo/app/chat-pull-hook.hoon:20]: Please remove bol and just call this bowl. Removes possibility of error/confusion in calling `bowl` directly.

File: pkg/arvo/app/chat-pull-hook.hoon [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+::
+++  on-poke
+  |=  [=mark =vase]
+  ^-  (quip card _this)
+  ?.  ?=(%chat-pull-hook-action mark)
+    (on-poke:def mark vase)
+  =/  act  !<(action:hook vase)
+  ?-  -.act
+      %add
+    ?>  (team:title our.bol src.bol)
```
Comment (loganallenc) [inline pkg/arvo/app/chat-pull-hook.hoon:48]: Please move this prior to the `?-` as it should apply to all conditions.

File: pkg/arvo/app/chat-pull-hook.hoon [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+            ~[/tracking]
+            %chat-pull-hook-update
+            !>([%tracking tracking])
+    ==  ==
+  ::
+      %remove
+    ::  XX chat-hook had a path for reading the ship out of the path if not
+    ::  present in the tracking map. Why is this helpful or necessary?
+    ::  If it's not in the map, what's the effect of removing it?
+    ::  Canceling subscriptions that shouldn't exist?
```
Comment (loganallenc) [inline pkg/arvo/app/chat-pull-hook.hoon:73]: Yes, the idea was to allow cancellation of subscriptions that stuck around when they shouldn’t have. Feel free to ignore that code path and not write it.

File: pkg/arvo/app/chat-pull-hook.hoon [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+      ?.  (~(has by tracking) chat)  [~ this]
+      =/  =ship
+        ?:  =('~' i.t.wire)
+          (slav %p i.t.t.wire)
+        (slav %p i.t.wire)
+      =/  =path  ?~((chat-scry:lib bol chat) wire [%mailbox chat])
+      :_  this
+      [%pass path %agent [ship %chat-push-hook] %watch path]~
+    ==
+
```
Comment (loganallenc) [inline pkg/arvo/app/chat-pull-hook.hoon:147]: Blank line

File: pkg/arvo/app/chat-pull-hook.hoon [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        ?~  shp  ~
+        ?.  =(src.bol u.shp)  ~
+        [(chat-poke [%messages path.update envelopes.update])]~
+      ==
+      ::
+      ++  chat-poke
+        |=  act=action:store
+        ^-  card
+        [%pass / %agent [our.bol %chat-store] %poke %chat-action !>(act)]
+      --
```
Comment (loganallenc) [inline pkg/arvo/app/chat-pull-hook.hoon:232]: Dedent and put a `::` on the next line

File: pkg/arvo/app/chat-pull-hook.hoon [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+                %agent
+                [our.bol %chat-view]
+                %poke
+                %chat-view-action
+                !>([%join shp app-path ask-history])
+        ==  ==
+      ==
+    --
+  ==
+  ::
```
Comment (loganallenc) [inline pkg/arvo/app/chat-pull-hook.hoon:253]: Unnecessary

File: pkg/arvo/app/chat-pull-hook.hoon [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+                /
+                %agent
+                [our.bol %chat-view]
+                %poke
+                %chat-view-action
+                !>([%join shp app-path ask-history])
+        ==  ==
+      ==
+    --
+  ==
```
Comment (loganallenc) [inline pkg/arvo/app/chat-pull-hook.hoon:252]: This should be on the same column as the opening `|^`. Could be mobile formatting but it doesn’t look like it is.

File: pkg/arvo/app/chat-push-hook.hoon [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+        %json  (chat-action (action:dejs:store !<(json vase)))
+        %chat-action  (chat-action !<(action:store vase))
+        %chat-push-hook-action  (chat-push-hook-action !<(action:hook vase))
+    ==
+  [cards this]
+  ::
+  ++  chat-action
+    |=  act=action:store
+    ^-  (quip card _state)
+    ::  XX why do we use ?> here but ?. ~ below?
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:53]: The thought is that we want it to loudly squawk at us if it gets unexpected input, but quietly decline if someone sends a message they’re not permitted to send.

File: pkg/arvo/app/chat-push-hook.hoon [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+    ?>  ?=(%message -.act)
+    :_  state
+    ?.  (~(has in sharing) path.act)  ~
+    ?.  (is-permitted:lib bol src.bol path.act)  ~
+    ::  XX do we still need this part?
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:58]: Yes, for better or for worse

File: pkg/arvo/app/chat-push-hook.hoon [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+            ?=(%code -.letter)
+            ?=(~ output.letter)
+        ==
+      =/  =hoon  (ream expression.letter)
+      letter(output (eval:store bol hoon))
+    =:  author.envelope.act  src.bol
+        when.envelope.act  now.bol
+    ==
+    [%pass / %agent [our.bol %chat-store] %poke %chat-action !>(act)]~
+
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:71]: Blank line

File: pkg/arvo/app/chat-push-hook.hoon [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+    ==
+    [%pass / %agent [our.bol %chat-store] %poke %chat-action !>(act)]~
+   
+  ::
+  ++  chat-push-hook-action
+    |=  act=action:hook
+    ^-  (quip card _state)
+    ?-  -.act
+        %add
+      ?>  (team:title our.bol src.bol)
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:78]: Just move this statement to before the switch

File: pkg/arvo/app/chat-push-hook.hoon [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+      =/  chat-wire  [%store path.act]
+      ?:  (~(has in sharing) path.act)  [~ state]
+      =:  sharing  (~(put in sharing) path.act)
+          allow-history  (~(put by allow-history) path.act allow-history.act)
+      ==
+      :_  state
+      [%pass chat-wire %agent [our.bol %chat-store] %watch chat-path]~
+    ::
+        %remove
+      ?.  !(team:title our.bol src.bol)
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:89]: Unnecessary

File: pkg/arvo/app/chat-push-hook.hoon [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+        [~ state]
+      ?.  (~(has in sharing) path.act)
+        ~&  [dap.bol %already-not-sharing path.act]
+        [~ state]
+      =.  sharing  (~(del in sharing) path.act)
+      :_  state
+      :~  [%give %kick [%mailbox path.act]~ ~]
+          [%pass [%mailbox path.act] %agent [our.bol %chat-store] %leave ~]
+          ::  XX is it worth kicking backlog subscriptions here just to be
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:98]: Yes, good idea. You’d have to do a similar iteration over the current ones to what is written in the pull-hook though

File: pkg/arvo/app/chat-push-hook.hoon [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+    ?~  tang  [~ this]
+    ?.  ?=([%store *] wire)  [~ this]
+    =/  pok  [%poke %chat-push-hook-action !>([%remove t.wire])]
+    :_  this
+    [%pass / %agent [our.bol %chat-push-hook] pok]~
+  ::
+  ++  fact-chat-update
+    |=  =update:store
+    ^-  (quip card _this)
+    ?>  (team:title our.bol src.bol)  ::  XX or do I need to use ?.  [~ this]
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:190]: Good as is

File: pkg/arvo/lib/chat-hooks.hoon [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+      bol
+      %permission-store
+      [%permitted (scot %p who) group-path]
+  ==
+::
+++  chats-of-group
+  |=  [bol=bowl:gall =group-path]
+  ^-  (list path)
+  ::  if metadata-store isn't running yet, we're still in the upgrade ota phase.
+  ::  we can't get chats from the metadata-store, but can make assumptions
```
Comment (loganallenc) [inline pkg/arvo/lib/chat-hooks.hoon:31]: These comments can be removed

File: pkg/arvo/lib/chat-hooks.hoon [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+  ^-  (unit path)
+  ?.  =(%chat app-name)  ~
+  `app-path
+::
+++  groups-of-chat
+  |=  [bol=bowl:gall chat=path]
+  ^-  (list group-path)
+  ::  if metadata-store isn't running yet, we're still in the upgrade ota phase.
+  ::  we can't get groups from the metadata-store, but can make assumptions
+  ::  about chat path shape, and the chat that would match it.
```
Comment (loganallenc) [inline pkg/arvo/lib/chat-hooks.hoon:60]: Remove plz

File: pkg/arvo/lib/chat.hoon (hunk 1 above)
Comment (loganallenc) [reply pkg/arvo/lib/chat.hoon:3]: /lib/userspace 👍🏻

File: pkg/arvo/mar/chat/pull-hook-update.hoon [hunk 19]
```diff
@@ -0,0 +1,7 @@
+/+  *chat-pull-hook
+|_  upd=update
+++  grab
+  |%
+  ++  noun  update
```
Comment (loganallenc) [inline pkg/arvo/mar/chat/pull-hook-update.hoon:5]: We need a json update and action 😉

**Comment by loganallenc** [comment]
Agreed that `=/` is fine without a type. From my perspective, the auto-naming is righteous, and three letter names should be reserved for variables that never leave the context of a small gate they’re being used within.

File: pkg/arvo/app/chat-pull-hook.hoon (hunk 3 above)
Comment (ohAitch) [reply pkg/arvo/app/chat-pull-hook.hoon:20]: Isn't the whole problem with autonaming error/confusion with calling `bowl` directly?

Comment (Fang-) [reply pkg/arvo/app/chat-pull-hook.hoon:20]: @ohAitch I think @loganallenc means `=bowl:gall`, equivalent to `bowl=bowl:gall`? We use this all over the place these days, and it's not a problem since you generally don't refer to the bowl type elsewhere, and it'd still be at `bowl:gall` if you needed it.

File: pkg/arvo/app/chat-push-hook.hoon (hunk 10 above)
Comment (Fang-) [reply pkg/arvo/app/chat-push-hook.hoon:53]: Don't you want it to be loud in both cases? This lets the sender/client know that it failed, instead of just wondering why they aren't getting their messages echoed back to them despite getting an ack on the poke. Not sure what you gain by hiding that.

**Comment by matildepark** [comment]
Happy to play the fool here, so what's the rationale for this PR? (And should this really be targeting `master`?) 

We really need to get out of the habit of these totally silent PR descriptions.

File: pkg/arvo/app/chat-pull-hook.hoon (hunk 4 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-pull-hook.hoon:48]: I would like that behavior to change, yes. No need for the host to be able to send a remove.

File: pkg/arvo/app/chat-pull-hook.hoon (hunk 2 above)
Comment (loganallenc) [reply pkg/arvo/app/chat-pull-hook.hoon:9]: Noted. Thanks!

File: pkg/arvo/mar/chat/pull-hook-update.hoon (hunk 19 above)
Comment (loganallenc) [reply pkg/arvo/mar/chat/pull-hook-update.hoon:5]: Yes.

**Comment by pilfer-pandex** [comment]
Test plan (using chat-cli):

- launch zod, nec, bus
- create a channel on zod
- using chat-cli, send a message from zod to the channel
  - verify that everyone has it
- using chat-cli, send a message from nec to the channel
  - verify that everyone has it
- using chat-cli, send (add 2 2) from zod to the channel
  - verify that everyone has it, evaluated
- using chat-cli, send send (add 2 2) from nec to the channel
  - verify that everyone has it, evaluated
- join from bus
  - verify that we get backlog [but see NOTE 1 below]
- send a message from bus to the channel
  - verify that everyone has it
- leave from nec, rejoin [broken before my change: in chat-cli, if you leave a channel, it is still listed in the state. so you don't get backlog but can still send messages to it.]
  - verify that we get backlog
- send (add 2 2) from bus to the channel
  - verify that everyone has it, evaluated
- on zod, kick nec from the channel, using the web interface
- on zod, delete the channel

[NOTE 1]
NOTE: ames error, even though success:
crud: %ames-wake event failed
bail: exit
wire=/pump/~zod/4
bar-stack=~[~[/ames] ~[//behn]]
take: failed
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[613 3].[633 20]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[614 3].[633 20]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[618 3].[633 20]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[620 3].[633 20]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[621 5].[631 7]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[622 5].[631 7]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[626 21].[626 62]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.052 5].[1.066 69]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.054 5].[1.066 69]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.055 5].[1.066 69]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.059 5].[1.066 69]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.060 5].[1.066 69]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.064 5].[1.066 69]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.066 5].[1.066 69]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.066 10].[1.066 69]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.658 7].[1.716 38]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.664 7].[1.716 38]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.677 7].[1.716 38]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.693 7].[1.716 38]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.709 7].[1.716 38]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.716 7].[1.716 38]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.747 7].[1.821 9]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.750 7].[1.821 9]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.753 7].[1.821 9]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.754 7].[1.821 9]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.754 47].[1.754 71]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.983 5].[1.990 7]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.985 5].[1.990 7]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.985 9].[1.985 29]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.995 5].[2.006 7]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.997 5].[2.006 7]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[1.999 14].[1.999 51]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[2.141 5].[2.146 50]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[2.143 5].[2.146 50]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[2.144 7].[2.144 42]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[2.194 5].[2.203 7]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[2.196 5].[2.203 7]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[2.198 5].[2.203 7]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[2.201 14].[2.201 36]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[2.208 5].[2.248 50]>
/~zod/home/~2020.6.4..02.18.34..d667/sys/vane/ames:<[2.211 5].[2.248 50]>

**Comment by philipcmonk** [comment]
That looks like the bug fixed in #2947.  I assume you last merged from master last week, if so you can fix this by merging from master again.

**Comment by pilfer-pandex** [comment]
Test plan for the migration:

- start zod and nec with code from before this change
- create a channel on zod and send some messages
- join the channel on nec and send some messages
- create a new channel on nec and send some messages
- update and commit the new code on zod
- update and commit the new code on nec
- send some messages to zod's channel from zod
- send some messages to zod's channel from nec
- join nec's channel from zod and send some messages

File: pkg/arvo/app/chat-push-hook.hoon [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+  ?+  path          (on-watch:def path)
+      [%backlog *]  [(backlog t.path) this]
+      [%mailbox *]  [(mailbox t.path) this]
+  ==
+  ::
+  ++  provider-for
+    |=  [=ship =rid]
+    ^-  ^ship
+    =/  candidates  ~(tap in (~(got by providers) rid))
+    =/  idx  (mod ship (lent `(list @p)`candidates))
```
Comment (philipcmonk) [inline pkg/arvo/app/chat-push-hook.hoon:153]: Galaxies are the low bits, so for example descendants of `~zod` are all `0 mod 256`.  Since most active ships are descendants of a a small number of galaxies, better to do `(mug ship)` or something.

File: pkg/arvo/app/chat-hook.hoon [hunk 21]
```diff
@@ -33,714 +32,69 @@
       invite-created=_|
       allow-history=(map path ?)
   ==
-::
-+$  poke
-  $%  [%chat-action action:store]
-      [%permission-action permission-action]
-      [%invite-action invite-action]
-      [%chat-view-action action:view]
```
Comment (loganallenc) [inline pkg/arvo/app/chat-hook.hoon:41]: This file cannot be deleted because we do not have app tombstoning yet. It could instead be replaced with a stub that deletes its own state and kills all of its subscriptions. I'm also curious how you're handling the OTA case if this file has been deleted?

File: pkg/arvo/app/chat-push-hook.hoon [hunk 22]
```diff
@@ -0,0 +1,339 @@
...
+    ?~  box  !!
+    [%give %fact ~ %chat-update !>([%create path])]~
+  --
+::
+++  on-agent
+  |=  [=wire =sign:agent:gall]
+  ^-  (quip card _this)
+  |^
+  ?+  -.sign                  (on-agent:def wire sign)
+      %kick  ?+  wire         !!
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:222]: Weird indentation. Backdent queenside as follows:

```
    %kick
  ?+  wire  !!
      [%permissions ~]  kick-permissions
  ==
::
```

File: pkg/arvo/app/chat-push-hook.hoon [hunk 23]
```diff
@@ -0,0 +1,339 @@
...
+          [%permissions ~]    kick-permissions
+          [%store @ *]        (kick-store wire)
+      ==
+      %watch-ack              (watch-ack wire p.sign)
+      %fact  ?+  p.cage.sign  (on-agent:def wire sign)
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:227]: Same.

File: pkg/arvo/app/chat-push-hook.hoon [hunk 24]
```diff
@@ -0,0 +1,339 @@
...
+      %+  turn
+        (~(app-paths-from-group metadata bowl) %chat path)
+      |=  chat=^path
+      ^-  (list card)
+      ?.  (~(has by providers) (path-to-rid chat))  ~
+      %-  zing
+      %+  turn  ~(tap in who)
+      |=  =ship
+      ?:  (is-permitted bowl ship chat)
+        ::  XX If they are permitted, under what cicumstances will it be
```
Comment (loganallenc) [inline pkg/arvo/app/chat-push-hook.hoon:298]: If they are permitted, they don't ever need to be removed. But notice that the case in which we do nothing on 300 is when they were previously on a blacklist and we have removed them from that blacklist.

File: pkg/arvo/sur/chat-pull-hook.hoon [hunk 25]
```diff
@@ -0,0 +1,19 @@
+/+  *userspace
+^?
+|%
++$  action
+  $%  ::  %add: subscribe to a foreign chatroom, mirroring its contents into
+      ::  our store
+      ::
+      [%add =ship =path ask-history=?]
```
Comment (loganallenc) [inline pkg/arvo/sur/chat-pull-hook.hoon:8]: Why is this taking in a path when we have resource identifiers now?

File: pkg/arvo/sur/chat-push-hook.hoon [hunk 26]
```diff
@@ -0,0 +1,24 @@
+^?
+|%
++$  action
+  $%  ::  %add: make a local chatroom accessible to foreign ships
+      ::
+      [%add =path allow-history=?]
```
Comment (loganallenc) [inline pkg/arvo/sur/chat-push-hook.hoon:6]: Same, this should be moved to a resource identifier, not a path.

**Comment by loganallenc** [comment]
Please rebase on top of Liam's lf/group-refactor, and target release/next-userspace instead of master.

File: pkg/arvo/app/chat-push-hook.hoon (hunk 24 above)
Comment (pilfer-pandex) [reply pkg/arvo/app/chat-push-hook.hoon:298]: So if they're being removed from a blacklist, that's a %remove fact as well? Got it.

File: pkg/arvo/app/chat-hook.hoon (hunk 21 above)
Comment (pilfer-pandex) [reply pkg/arvo/app/chat-hook.hoon:41]: Not sure what you're getting at here; the file *isn't* being deleted. I've basically done what you've asked: left a stub which deletes subscriptions. It doesn't kill the state because I need that around scryable so that when the new hooks start up, they can move over the state (e.g. the fact that you're subscribed to urbit-help or whatever).

File: pkg/arvo/sur/chat-push-hook.hoon (hunk 26 above)
Comment (pilfer-pandex) [reply pkg/arvo/sur/chat-push-hook.hoon]: Good point.

**Comment by pilfer-pandex** [comment]
Merge authority: I am waiting for logan's review and any further back and forth before conducting tests. Please refrain from merging until I have certified the code. Thanks.

**Comment by loganallenc** [comment]
Could you change this to a draft PR? I’d like to approve it and take it off your hands but not necessarily merge it until I’m back online and can talk to Ted and Matilde about an integration plan.

**Comment by pilfer-pandex** [comment]
Any steps I should take from here, or ways I can help with testing after js
changes? Note that channels-with-groups haven't been tested since I
couldn't figure out how to do that with chat-cli (and was told it wasn't
possible).

—
~pilfer-pandex

**Comment by loganallenc** [comment]
I can take it from here ☺️
